compiler/p1/doc/p1red.man
.TH WhiteSmith 1 "" "WhiteSmith"
.SH NAME
p1red.c \- reduces expressions
.SH SYNOPSIS
p1red.c performs compile time expression reduction
and machine independent tree rewriting
.LP
recodes is a table of deductions which is mapped
onto the operators found in p1type.c.
The codes represented are:
.EX
	0200	reduce right side of the expression
	0100	reduce left side of the expression
	0040	if left side is a term commute left and right
	0020	if right side is zero make the expression left side, exit
	0010	if right side is zero make the expression zero, exit
	0004	if right side is one make the expression left side, exit
	0002	if right side isn't constant exit
	0001	if left side isn't constant exit
.EE
.LP
recodes is used by function reduce.
The order of applying the codes to an expression is important.
An example:
.EX
	LTIMES is 0354
	1.	reduce right
	2.	reduce left
	3.	if left term commute
	4.	if right zero, expr is zero, exit
	5.	if right one, expr is one, exit
.EE
.LP
canadd does compile time addition and returns whether it could.
If both sides of the plus operator are integer,
have no operator and no refcount,
and either side has no name and no register, then a compile time add is done.
To do compile time addition, name, bias and register (idx) are added into the left side;
and the maximum type is set up to LONG (with PTRTO being considered less then LONG).
.LP
canmul does compile time multiplication on pure constants.
The type is set to maximum of long (not including PTRTO).
.LP
cansub does compile time subtraction.
If both left and right sides have the same name, the subtraction is done,
with the new name being null.
Also, if both left and right have the same idx,
idx is set to zero.
Type need not be set since either the right has only a bias
or has the same name and/or idx as left.
.LP
reduce does the expression reduction, using the recodes tables.
First, for any expression, reduce applies the recodes table entry
and does one or more of the eight reductions.
Then, and operator specific reduction is done;
note that for the switch controlling this, break causes
q to be set to l.
Among the reductions are:
.EX
	address of:		If left is a term, reduce ref count by one.
					If applied to indirection, remove both address of
					and indirection.
	cast:			Redundant casts are dropped.
	indirect:		if left is a term, add one to refcount.
					Thus for *i, i has a refcount of two.
					Else if left op is address of,
					remove both indirect and address of.
	query (?):		Reduce the middle term of the ternary.
					and if the test (left side) is constant,
					choose right or middle at complile time.
	plus:			If canadd left to right then copy up left.
					else regroup the tree if possible.
	minus:			If can subtract left from right,
					subtract and copy up left,
					else if right is a constant regroup the tree.
	times:			If can multiply, do it and copy up left,
					else regroup the tree.
.EE


compiler/p1/p1red.c
/*	EXPRESSION REDUCTION
 *	copyright (c) 1978 by Whitesmiths, Ltd.
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "int1.h"
#include "int01.h"
#include "int12.h"
#include "int012.h"
#include "util.h"
#include "p1.h"

/*	the reducing tables
	0200	reduce right
	0100	reduce left
	0040	if (L is a term) commute
	0020	if (R is zero) take as L and exit
	0010	if (R is zero) take as 0 and exit
	0004	if (R is one) take as L and exit
	0002	if (R not const) exit
	0001	if (L not const) exit
 */
static TINY recodes[] = {
	0, 0100, 0100, 0, 0101, 0100, 0101,
	0101, 0100, 0100, 0300, 0300, 0100,
	0300, 0300, 0300, 0300, 0300, 0300, 0300, 0300,
	0300, 0300, 0300, 0300, 0300, 0360, 0320,
	0354, 0307, 0303, 0353, 0363, 0363, 0323, 0323,
	0303, 0303, 0303, 0303, 0343, 0343, 0317, 0323, 0, 0
};

/*
 * check if can do arithmetic
 */
static BOOL cachk(TERM *p)
{
	return !p->op && !p->e.v.refs && p->ty != TFLOAT && p->ty != TDOUBLE;
}


/*
 * add if you can
 */
static BOOL canadd(TERM *l, TERM *r)
{
	if (cachk(l) && cachk(r) && (!l->n.an[0] || !r->n.an[0]) && (!l->e.v.idx || !r->e.v.idx))
	{
		if (!l->n.an[0])
			cpynm(l->n.an, r->n.an);
		l->e.v.bias += r->e.v.bias;
		l->e.v.idx += r->e.v.idx;
		if (type(l->ty) == TPTRTO)
			;
		else if (type(r->ty) == TPTRTO)
		{
			l->ty = r->ty;
			l->at = r->at;
		} else
			l->ty = maxty(l->ty, r->ty, TULONG);
		return TRUE;
	}
	return FALSE;
}


/*
 * multiply if you can
 */
static BOOL canmul(TERM *l, TERM *r)
{
	if (iscons(l) && iscons(r))
	{
		l->e.v.bias *= r->e.v.bias;
		l->ty = maxty(l->ty, r->ty, TULONG);
		return TRUE;
	}
	return FALSE;
}


/*
 * subtract if you can
 */
static BOOL cansub(TERM *l, TERM *r)
{
	if (cachk(l) && cachk(r) &&
		(!r->n.an[0] || memcmp(l->n.an, r->n.an, LENNAME) == 0) && (!r->e.v.idx || l->e.v.idx == r->e.v.idx))
	{
		if (r->n.an[0])
			cpynm(l->n.an, noname);
		if (r->e.v.idx)
			l->e.v.idx = 0;
		l->e.v.bias -= r->e.v.bias;
		return TRUE;
	}
	return FALSE;
}


/*
 * copy a term
 */
static TERM *cpyterm(TERM *to, TERM *from)
{
	if (!from->op)
	{
		setad(to, from->n.an, from->e.v.bias, from->e.v.idx, from->e.v.refs);
	} else
	{
		to->op = from->op;
		to->e.o.left = from->e.o.left;
		to->e.o.right = from->e.o.right;
		to->e.o.mid = from->e.o.mid;
	}
	setty((SYMBOL *)to, from->ty, from->at);
	return to;
}


/*
 * reduce an expression
 */
TERM *reduce(TERM *q)
{
	TERM *l;
	TERM *r;
	int code;

	if (!q)
		return NULL;
	l = q->e.o.left;
	r = q->e.o.right;
	code = recodes[_scnstr(tyops, q->op)];
	if (code & 0200)
		reduce(r);
	if (code & 0100)
		reduce(l);
	if (code & 0040)
	{
		if (!l->op && (r->op || iscons(l)))
		{
			q->e.o.left = r;
			q->e.o.right = l;
			l = r;
			r = q->e.o.right;
		}
	}
	if (code & 0020)
		if (iscons(r) && r->e.v.bias == 0)
			return cpyterm(q, l);
	if (code & 0010)
		if (iscons(r) && r->e.v.bias == 0)
			return cpyterm(q, r);
	if (code & 0004)
		if (iscons(r) && r->e.v.bias == 1)
			return cpyterm(q, l);
	if (code & 0002)
		if (!iscons(r))
			return q;
	if (code & 0001)
		if (!iscons(l))
			return q;
	switch (q->op)
	{
	case DADDR:
		if (!l->op)
		{
			--l->e.v.refs;
		} else if (l->op == DINDIR)
		{
			cpyterm(q, l->e.o.left);
			return q;
		} else
		{
			return q;
		}
		break;
	case DCAST:
		if (!l->op && (type(q->ty) == type(l->ty) || (!dtype(q->ty) && !dtype(l->ty))))
			;
		else
			return q;
		break;
	case DINDIR:
		if (!l->op)
		{
			++l->e.v.refs;
		} else if (l->op == DADDR)
		{
			cpyterm(q, l->e.o.left);
			return q;
		} else
		{
			return q;
		}
		break;
	case LNOT:
		l->e.v.bias = !l->e.v.bias;
		break;
	case LCOMP:
		l->e.v.bias = ~l->e.v.bias;
		break;
	case DMINUS:
		if (dlit(l))
			l->n.dn = -l->n.dn;
		else if (iscons(l))
			l->e.v.bias = -l->e.v.bias;
		else
			return q;
		break;
	case DPLUS:
		if (dlit(l) || iscons(l))
			break;
		else
			return q;
		break;
	case LQUERY:
		reduce(q->e.o.mid);
		if (iscons(l))
			cpyterm(q, l->e.v.bias ? (TERM *) q->e.o.mid : r);
		return q;
	case LPLUS:
		if (canadd(l, r))
		{
			break;
		} else if (l->op == LPLUS && !dtype(q->ty) && canadd(l->e.o.right, r))
		{
			l->ty = q->ty;
			l->at = q->at;
			return cpyterm(q, l);
		} else
		{
			return q;
		}
	case LMINUS:
		if (cansub(l, r))
		{
			break;
		} else if (iscons(r))
		{
			q->op = LPLUS;
			r->e.v.bias = -r->e.v.bias;
			return q;
		} else
		{
			return q;
		}
	case LTIMES:
		if (canmul(l, r))
		{
			break;
		} else if (l->op == LPLUS && !dtype(q->ty) && canmul(l->e.o.right, r))
		{
			l->op = LTIMES;
			q->op = LPLUS;
			q->e.o.right = l->e.o.right;
			l->e.o.right = r;
			return q;
		} else
		{
			return q;
		}
	case LDIVIDE:
		if (!r->e.v.bias)
			p1error("illegal /");
		else
			l->e.v.bias /= r->e.v.bias;
		break;
	case LMODULO:
		if (!r->e.v.bias)
			p1error("illegal %");
		else
			l->e.v.bias %= r->e.v.bias;
		break;
	case LAND:
		l->e.v.bias &= r->e.v.bias;
		break;
	case LOR:
		l->e.v.bias |= r->e.v.bias;
		break;
	case LXOR:
		l->e.v.bias ^= r->e.v.bias;
		break;
	case LLSHIFT:
		l->e.v.bias <<= r->e.v.bias;
		break;
	case LRSHIFT:
		l->e.v.bias >>= r->e.v.bias;
		break;
	case LLESS:
		l->e.v.bias = (l->e.v.bias < r->e.v.bias);
		break;
	case LLEQ:
		l->e.v.bias = (l->e.v.bias <= r->e.v.bias);
		break;
	case LGREAT:
		l->e.v.bias = (l->e.v.bias > r->e.v.bias);
		break;
	case LGEQ:
		l->e.v.bias = (l->e.v.bias >= r->e.v.bias);
		break;
	case LISEQ:
		l->e.v.bias = (l->e.v.bias == r->e.v.bias);
		break;
	case LNOTEQ:
		l->e.v.bias = (l->e.v.bias != r->e.v.bias);
		break;
	default:
		return q;
	}
	setad(q, l->n.an, l->e.v.bias, l->e.v.idx, l->e.v.refs);
	return q;
}
